# 10. Data analysis

It is time to combine everything we have learned so far, to do a data analysis - an essential task for data science.

Here is the table of contents for this notebook:

- 10.1 Describing the dataset
- 10.2 Calculating the mean of the value column
- 10.3 Improving the implementation
- 10.4 Exercises

## 10.1 Describing the dataset

Let's `open` and `print` the all the lines of `data.csv`:

In [ ]:
fhand = open('data.csv')
for line in fhand:
    print(line)
fhand.close()

This dataset is part of the Sustainable Development Goals (SDG). There are 4 columns. The first column is `indicator`. This indicator is part of **Goal 7**:

>Ensure access to affordable, reliable, sustainable and modern energy for all

and **indicator 7.1.2** is the:

>Proportion of population with primary reliance on clean fuels and technologies for cooking (%)

second column is `geoareaname` which is short for Geographical Area Name. The `timeperiod` shows when the indicator is measured and `value` column is the value of the indicator in percentages. Values higher than 95% are shown as `>95`. Also there are values less than 5% represented by `<5`. For our analysis we will assume >95% to be 95% and <5% to be 5%.

## 10.2 Calculating the mean of the value column

Let's find the mean of the `value` column.

But all we have is the whole line, how can we access the value? Let's look at the second line (first line is column names) to figure out how we can extract the value.

In [1]:
fhand = open('data.csv')
count = 0
for line in fhand:
    print(line)
    count = count + 1
    if count == 2:
        break
fhand.close()

indicator;geoareaname;timeperiod;value

7.1.2;Bhutan;2015;52



How can we extract the number 52 from this line?

In [2]:
line

'7.1.2;Bhutan;2015;52\n'

We know that the values are at the end so perhaps we can slice the line.

**Exercise 10.1**

Slice the line to get `52` from `line`

In [4]:
for datapoint in line.split(';'):
    print(datapoint)

7.1.2
Bhutan
2015
52



Hopefully you managed to do it. But we want an approach that works for _all_ lines. So let's _test_ it for all lines:

In [5]:
fhand = open('data.csv')
for line in fhand:
    print(line[-3:-1]) # Exercise 10.1
fhand.close()

ue
52
95
95
95
95
95
54
;7
95
79
25
17
92
<5
14
95
<5
41
94
95
95
;9
79
74
95
51
90
85
89
95
91
95
<5
65
<5
95
95
35
95
<5
95
36
95
95
59
<5
95
94
;7
14
12
95
95
95
58
68
44
95
95
87
<5
24
91
46
50
79
95
<5
12
95
<5
95
90
<5
57
13
86
81
95
95
63
92
65
53
90
95
95
40
66
;4


It seems like we have a problem 😱 Our genious approach that worked for a single case, did not generalize to all cases 😢

It worked for most cases, but that is not enough. Let's see the cases where it fails:

- Case 1: First line is column names, so it does not contain a value
- Case 2: Single digit numbers (`;7`, `<5`)

When we designed our algorithm, our example line had a 2-digit number. But as we have just seen that is not always the case. You will encounter this concept over and over again: when you are programming, you have to consider possible cases as much as you can.

Considering all cases when writing a program in Python is essential for producing reliable, efficient, and bug-free software. By covering a wide range of scenarios, including normal, boundary, and _edge cases_, developers can enhance the program's functionality, handle unexpected situations gracefully, and provide a better user experience. Thorough testing and validation with various input values help identify and resolve potential issues, ensuring that the program performs as expected in any given situation.

It is hard to foresee all the cases, therefore _testing_ your code is key. In fact, that is how we discovered the error in our code above.

- We developed an approach for a single case
- Tested it in all lines
- Detected failure in some cases.

### 🐍 Advanced 🐍

This is called _unit testing_. It is a _unit_, because we tested a small piece (extracting a value from a line) of the source code. You can test your code in many different levels from testing units to testing systems. We are not going to cover unit testing in this block. If you would like to learn more, take a look at the following tutorial:

[Python Tutorial: Unit Testing Your Code with the unittest Module](https://www.youtube.com/watch?v=6tNS--WetLI)

Now, let's continue and come up with an approach that would work for all cases.

We can handle the first case by skipping the first line, as follows:

In [ ]:
fhand = open('data.csv')
count = 0
for line in fhand:
    # Skip the first line
    if count == 0:
        count += 1
        continue

    print(line[-3:-1])
    count = count + 1

fhand.close()

For the second case, we can check if the string can be converted to an integer or not and catch these cases with `try`/`except`

In [ ]:
int(';7')

In [ ]:
try:
    int(';7')
    print('Double digit')
except:
    print('Single digit')

Let's add this logic into our program.

In [ ]:
fhand = open('data.csv')
count = 0
for line in fhand:
    # Skip the first line
    if count == 0:
        count += 1
        continue
    
    # Slice the number
    number = line[-3:-1]
    
    try:
        # For two digit numbers this should work
        # It will raise an exception for single digit numbers, except block will run
        number = int(number)
        print(number)
    except:
        # If this code block is executed
        # That means we have a single digit number
        # We can slice accordingly
        number = line[-2:-1]
        number = int(number)
        print(number)
    
    count = count + 1

fhand.close()

It worked 🥳 To calculate the mean of the column, we need to save these numbers into a data structure (e.g. a list) and calculate their mean.

In [ ]:
fhand = open('data.csv')
numbers = [] # create an empty list
count = 0
for line in fhand:
    # Skip the first line
    if count == 0:
        count += 1
        continue
    
    # Slice the number
    number = line[-3:-1]
    
    try:
        # For two digit numbers this should work
        # It will raise an exception for single digit numbers, except block will run
        number = int(number)
        numbers.append(number) # append to list instead of printing
    except:
        # If this code block is executed
        # That means we have a single digit number
        # We can slice accordingly
        number = line[-2:-1]
        number = int(number)
        numbers.append(number) # append to list instead of printing
print(sum(numbers)/len(numbers))
fhand.close()

That's it! 

Average of the proportion of population with primary reliance on clean fuels and technologies for cooking is 63.5%.

## 10.3 Improving the implementation

Let's remember the Zen of Python:

In [6]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


Let's see if we can make our code more _Pythonic_.

>Exploiting the features of the Python language to produce code that is clear, concise and maintainable. Pythonic means code that doesn't just get the syntax right, but that follows the conventions of the Python community and uses the language in the way it is intended to be used. [Source](https://stackoverflow.com/a/25011492).

Let's look at the first line again.

In [ ]:
fhand = open('data.csv')
count = 0
for line in fhand:
    print(line)
    count = count + 1
    if count == 2:
        break
fhand.close()

In [ ]:
line

Whenever you are working with strings, try to remember and use the string methods. We have learned about `strip` and `split` methods.

**Exercise 10.2**

Use the `strip` method to get rid of the newline character `\n`

In [11]:
line = '7.1.2;Bhutan;2015;52\n'
line.strip()

'7.1.2;Bhutan;2015;52'

**Exercise 10.3**

Use the `split` method to split the remaining string. Use the delimiter `;`.

The result should be
`['7.1.2', 'Bhutan', '2015', '52']`


In [10]:
line = '7.1.2;Bhutan;2015;52' # line after exercise 10.2

line.split(';')

['7.1.2', 'Bhutan', '2015', '52']

**Exercise 10.4**

Access the value we are looking for (i.e. 52) from the list `['7.1.2', 'Bhutan', '2015', '52']`

In [13]:
line = ['7.1.2', 'Bhutan', '2015', '52']
line[-1]

'52'

This might look more complex than slicing the string, but this approach will cover more cases. In other words, slicing assumed a two-digit number in a fixed location, this code will work regardless of the number of digits. Additionally, hardcoding a slice with indices (e.g. `line[-3:-1]`) is prone to slight differences in lines.

In [14]:
fhand = open('data.csv')
for line in fhand:
    line = line.strip() # Exercise 10.2
    line = line.split(';') # Exercise 10.3
    number = line[-1] # Exercise 10.4
    print(number)
fhand.close()

value
52
>95
>95
>95
>95
>95
54
7
>95
79
25
17
92
<5
14
>95
<5
41
94
>95
>95
9
79
74
>95
51
90
85
89
>95
91
>95
<5
65
<5
>95
>95
35
>95
<5
>95
36
>95
>95
59
<5
>95
94
7
14
12
>95
>95
>95
58
68
44
>95
>95
87
<5
24
91
46
50
79
>95
<5
12
>95
<5
>95
90
<5
57
13
86
81
>95
>95
63
92
65
53
90
>95
95
40
66
40


Now we still have the same problem with the first line, additionally we have to get rid of the `<` and `>` operators. We can use the `replace` method for that.

**Exercise 10.5**

Use the replace method to eliminate the greater operator (>).

In [15]:
number = '>95'
number.replace('>', '')

'95'

**Exercise 10.6**

Use the replace method to eliminate the smaller operator (<).

In [16]:
number = '<5'
number.replace('<', '')

'5'

Note that `replace` won't change the string if the logical operators are not in the string. So we can apply these to all numbers.

In [ ]:
fhand = open('data.csv')
for line in fhand:
    line = line.strip() # Exercise 10.2
    line = line.split(';') # Exercise 10.3
    number = line[-1] # Exercise 10.4
    number = number.replace('>', '') # Exercise 10.5
    number = number.replace('<', '') # Exercise 10.6
    print(number)
fhand.close()

We can make the code still more concise by chaining methods

In [ ]:
fhand = open('data.csv')
for line in fhand:
    number = line.strip().split(';')[-1] # Exercises 10.2-4
    number = number.replace('>', '').replace('<', '') # Exercises 10.5-6
    print(number)
fhand.close()

let's add skipping the first line, by using a slightly different logic, instead of counting.

In [17]:
fhand = open('data.csv')
for line in fhand:
    number = line.strip().split(';')[-1] # Exercises 10.2-4
    if number == 'value':continue # new logic for skipping the first line
    number = number.replace('>', '').replace('<', '') # Exercises 10.5-6
    print(number)
fhand.close()

52
95
95
95
95
95
54
7
95
79
25
17
92
5
14
95
5
41
94
95
95
9
79
74
95
51
90
85
89
95
91
95
5
65
5
95
95
35
95
5
95
36
95
95
59
5
95
94
7
14
12
95
95
95
58
68
44
95
95
87
5
24
91
46
50
79
95
5
12
95
5
95
90
5
57
13
86
81
95
95
63
92
65
53
90
95
95
40
66
40


That's it! Finally we create a list, and calculate the average.

In [18]:
fhand = open('data.csv')
numbers = []
for line in fhand:
    number = line.strip().split(';')[-1] # Exercises 10.2-4
    if number == 'value':continue # new logic for skipping the first line
    number = number.replace('>', '').replace('<', '') # Exercises 10.5-6
    numbers.append(int(number))
print(sum(numbers)/len(numbers))
fhand.close()

63.94444444444444


As you can see this code is much better than the previous version

```python
fhand = open('data.csv')
numbers = []
count = 0
for line in fhand:
    if count == 0:
        count += 1
        continue
    number = line[-3:-1]
    
    try:
    
        number = int(number)
        numbers.append(number)
    except:
        
        number = line[-2:-1]
        number = int(number)
        numbers.append(number)
print(sum(numbers)/len(numbers))
fhand.close()
```

This is nice but did you notice anything else? The previous result was 63.54% but with the second implementation we got 63.94% 😱😱😱😱

This is a major problem, we only changed our approach for extracting the values which shouldn't change the average. Discovering the issue is left for you as an exercise.

## 10.4 Exercises

**Exercise 10.7**

Investigate why these two implementations that are intended to calculate the same statistic, give different results. Determine which implementation is correct and which implementation is wrong.

In [19]:
# Implementation 1
fhand = open('data.csv')
numbers = []
count = 0
for line in fhand:
    if count == 0:
        count += 1
        continue
    number = line[-3:-1]
    try:
        number = int(number)
        numbers.append(number)
    except:
        number = line[-2:-1]
        number = int(number)
        numbers.append(number)
print(sum(numbers)/len(numbers))
fhand.close()

63.544444444444444


In [20]:
# Implementation 2
fhand = open('data.csv')
numbers = []
for line in fhand:
    number = line.strip().split(';')[-1] # Exercises 10.2-4
    if number == 'value':continue # new logic for skipping the first line
    number = number.replace('>', '').replace('<', '') # Exercises 10.5-6
    numbers.append(int(number))
print(sum(numbers)/len(numbers))
fhand.close()

63.94444444444444


**Exercise 10.8**

Fix the wrong implementation.

In [ ]:
# YOUR CODE HERE

**Exercise 10.9**

Find the geographical areas with proportion of population with primary reliance on clean fuels and technologies for cooking (%) less than or equal to 5 %.

Expected result

`['Rwanda',
 'Malawi',
 'Nigeria',
 'Guinea',
 'Niger',
 'Myanmar',
 'Sierra Leone',
 'South Sudan',
 'Madagascar',
 'Chad']`

In [50]:
file = open('data.csv')

cleanEnergyPercentagePerCountry = dict()
lowCleanEnergyCountries = []

for line in file:
    country = line.strip().split(';')[1]
    value = line.strip().split(';')[-1]

    if value == 'value':continue

    value = value.replace('<', '').replace('>', '')

    cleanEnergyPercentagePerCountry[country] = value

for country in cleanEnergyPercentagePerCountry:
    if int(cleanEnergyPercentagePerCountry[country]) <= 5:
        lowCleanEnergyCountries.append(country)

print(lowCleanEnergyCountries)
file.close()

['Myanmar', 'Rwanda', 'Malawi', 'Nigeria', 'Guinea', 'Niger', 'Sierra Leone', 'South Sudan', 'Madagascar', 'Chad']
